Miscellaneous RefCell cleanups

nealsid opened the pull request:
- Clarify `RefCell` error messages when borrow rules are broken
- Remove `Debug` impl for `BorrowError`/`BorrowMutError` since `#derive(Debug)` provides identical functionality
- Rename `BorrowFlag` to `BorrowCounter`
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

tgross35 commented:
Thank you for the PR!

The wording isn't quite accurate because `panic_already_borrowed` now says "...because it is also borrowed as **im**mutable", but `borrow_mut` will also fail if there is an existing mutable borrow. The current messages aren't ideal either because 

```rust
let cell = RefCell::new(0u32);
let b = (cell.borrow_mut(), cell.borrow_mut());
drop(b);
```

also prints "already borrowed: BorrowMutError" rather than the more accurate "already mutably borrowed: BorrowMutError", but it still works.

So I think the original wording of "already (mutably) borrowed (here: {err:?})" is probably fine to keep. We could alternatively add track mutably/immutably in the error type, but I don't expect that to be worth it since the `# Panics` API docs are pretty easy to locate (by comparison, with the rustc error it might not be as obvious that you are mutably/immutably borrowing because e.g. it's in a function signature).

> Fix the logging when debug_refcell is enabled so the location is output as a string rather than a sequence of bytes

Mind commenting the before and after here? `Location`'s debug should just print the file, line, and column.

tgross35 commented:
@rustbot author for the above

nealsid commented:
> 
> > Fix the logging when debug_refcell is enabled so the location is output as a string rather than a sequence of bytes
> 
> Mind commenting the before and after here? `Location`'s debug should just print the file, line, and column.

`Location` only implements `Display`, not `Debug` ([link](https://github.com/rust-lang/rust/blob/master/library/core/src/panic/location.rs#L205)).  Below is the output if you enable debug_refcell and cause a panic today:

```
     Running `target\debug\refcell-test.exe`
Hello, world! RefCell { value: 5 }

thread 'main' panicked at src\main.rs:11:15:
already borrowed: BorrowMutError { location: Location { file_bytes_with_nul: [115, 114, 99, 92, 109, 97, 105, 110, 46, 114, 115, 0], line: 8, col: 15 } }
stack backtrace:
   0: std::panicking::begin_panic_handler
             at C:\Users\neals\rust\library\std\src\panicking.rs:697
   1: core::panicking::panic_fmt
             at C:\Users\neals\rust\library\core\src\panicking.rs:75
   2: core::cell::panic_already_borrowed
             at C:\Users\neals\rust\library\core\src\cell.rs:791
   3: core::cell::RefCell<i32>::borrow_mut<i32>
             at C:\Users\neals\rust\library\core\src\cell.rs:1083
   4: refcell_test::main
             at .\src\main.rs:11
   5: core::ops::function::FnOnce::call_once<void (*)(),tuple$<> >
             at C:\Users\neals\rust\library\core\src\ops\function.rs:250
   6: core::hint::black_box
             at C:\Users\neals\rust\library\core\src\hint.rs:482
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
error: process didn't exit successfully: `target\debug\refcell-test.exe` (exit code: 101)
PS C:\Users\neals\refcell-test>
```

nealsid commented:
> Thank you for the PR!
> 
> The wording isn't quite accurate because `panic_already_borrowed` now says "...because it is also borrowed as **im**mutable", but `borrow_mut` will also fail if there is an existing mutable borrow. The current messages aren't ideal either because
> 
> ```rust
> let cell = RefCell::new(0u32);
> let b = (cell.borrow_mut(), cell.borrow_mut());
> drop(b);
> ```
> 
> also prints "already borrowed: BorrowMutError" rather than the more accurate "already mutably borrowed: BorrowMutError", but it still works.
> 
> So I think the original wording of "already (mutably) borrowed (here: {err:?})" is probably fine to keep. We could alternatively add track mutably/immutably in the error type, but I don't expect that to be worth it since the `# Panics` API docs are pretty easy to locate (by comparison, with the rustc error it might not be as obvious that you are mutably/immutably borrowing because e.g. it's in a function signature).

Sounds good, thanks!  I'll revert that part of the logging.

Your example code made me think of another potential issue:

```
    let v = RefCell::new(5);
    let w = v.borrow();
    let y = v.borrow();
    drop(w);
    let x = v.borrow_mut();
```
In this case, the `RefCell` will track the location of the borrow for `w` (since it tracks the first borrow: [link](https://github.com/rust-lang/rust/blob/master/library/core/src/cell.rs#L1024)) and log that in the panic, although `y` is the conflicting borrow when `borrow_mut()` is called on the last line.  Is that something that should be fixed?

tgross35 commented on library/core/src/cell.rs:
```diff
@@ -787,16 +787,24 @@ impl Display for BorrowMutError {
 #[cold]
-fn panic_already_borrowed(err: BorrowMutError) -> ! {
-    panic!("already borrowed: {:?}", err)
+fn panic_already_borrowed(_err: BorrowMutError) -> ! {
```
Right, but whether or not something gets used on all branches doesn't affect liveliness - one is enough. It would be different if the comptime `#[cfg(...)]` was used rather than the runtime `cfg!()` (semantically runtime at least, even though it's gone with the most basic optimizations) but there won't be any `dead_code` warnings here.

nealsid committed:
Two changes: Have BorrowError & BorrowMutError derive Debug and add
more information to Display implementation for BorrowError/BorrowMutError

- The BorrowError/BorrowMutError Debug implementations do not print
anything differently from what the derived implementation does, so we
don't need it.

- This change also adds the location field of
BorrowError/BorrowMutError to the the Display output when it is
present, rewords the error message, and uses the Display trait for
outputting the error message instead of Debug.

nealsid committed:
Rename BorrowFlag type to BorrowCounter

It's actually used as a counter so update the name to reflect that.

nealsid commented:
I seem to have hosed my repo, sorry about that! Please ignore.

nealsid commented:
@rustbot label +S-waiting-on-author -S-waiting-on-review

nealsid commented:
> The wording isn't quite accurate because `panic_already_borrowed` now says "...because it is also borrowed as **im**mutable", but `borrow_mut` will also fail if there is an existing mutable borrow. The current messages aren't ideal either because

Ah, I missed that, thanks!

> 
> ```rust
> let cell = RefCell::new(0u32);
> let b = (cell.borrow_mut(), cell.borrow_mut());
> drop(b);
> ```
> 
> also prints "already borrowed: BorrowMutError" rather than the more accurate "already mutably borrowed: BorrowMutError", but it still works.
> 
> So I think the original wording of "already (mutably) borrowed (here: {err:?})" is probably fine to keep. We could alternatively add track mutably/immutably in the error type, but I don't expect that to be worth it since the `# Panics` API docs are pretty easy to locate (by comparison, with the rustc error it might not be as obvious that you are mutably/immutably borrowing because e.g. it's in a function signature).
> 

Agreed on it not being worth it to track mutably/immutably in the error type.  However, I still think the error messages are a bit cryptic, even if the audience is developers (or, in the case of enabling `debug_refcell`, developers who have rebuilt std).  They require some spelunking into the code to understand, especially to understand the location field (because which previous borrow it picked to log isn't entirely clear).  

How about the following?

Without debug_refcell enabled:

```
thread 'main' panicked at src\main.rs:9:15:
RefCell already mutably borrowed
```

With debug_refcell enabled:

```
thread 'main' panicked at src\main.rs:9:15:
RefCell already mutably borrowed, a previous borrow was at this location: src\main.rs:8:15
```

nealsid commented:
@tgross35 In my testing I've noticed that the `Debug` impl for `BorrowError`/`BorrowMutError` don't output anything different from having `BorrowError`/`BorrowMutError` having `#[derive(Debug)]` added to them, unless I'm missing something.  What do you think about removing them? Should be less code to maintain while retaining backward-compatibility.

oli-obk commented on library/core/src/cell.rs:
```diff
@@ -12,7 +12,7 @@
 //! Shareable mutable containers exist to permit mutability in a controlled manner, even in the
 //! presence of aliasing. [`Cell<T>`], [`RefCell<T>`], and [`OnceCell<T>`] allow doing this in
-//! a single-threaded way—they do not implement [`Sync`]. (If you need to do aliasing and
+//! a single-threaded way-they do not implement [`Sync`]. (If you need to do aliasing and
```
This was intentional. A long dash renders differently from a normal dash/minus sign. After your change it looks like `way-they` is a word

tgross35 replied:
Tbh I think we should always recommend space-en-space, space-hyphen-hyphen-space, or even space-hyphen-space as opposed to em with no spaces. Not all style guides accept it, but those variants are reasonably common and it's actually understandable in monospace. Or else rustdoc should rewrite `--`/`---` (wouldn't mind that either).

tgross35 replied:
Apparently I've thought that for a while since I originally wrote it this way, mildly amused to find myself on the git log for this very specific dash 😆 https://github.com/rust-lang/rust/commit/9b5122976e9eb0f2f0b8def82fbcd972837880e4

tgross35 commented on library/core/src/cell.rs:
```diff
@@ -732,54 +732,52 @@ pub struct RefCell<T: ?Sized> {
-        builder.field("location", self.location);
+        let res = write!(
+            f,
+            "RefCell already mutably borrowed, a previous borrow was at this location: {}",
```
Two small tweaks: use `;` rather than `,` to join two independent clauses, and drop "at this location" since that's already reasonably clear.

```rust
            "RefCell already mutably borrowed; a previous borrow was at {}",
```

tgross35 commented on library/core/src/cell.rs:
```diff
@@ -788,15 +786,15 @@ impl Display for BorrowMutError {
 #[cold]
 fn panic_already_borrowed(err: BorrowMutError) -> ! {
-    panic!("already borrowed: {:?}", err)
+    panic!("{}", err)
 }
 
 // This ensures the panicking code is outlined from `borrow` for `RefCell`.
 #[cfg_attr(not(feature = "panic_immediate_abort"), inline(never))]
 #[track_caller]
 #[cold]
 fn panic_already_mutably_borrowed(err: BorrowError) -> ! {
-    panic!("already mutably borrowed: {:?}", err)
+    panic!("{}", err)
 }
```
Small nit, the inline formatting is usually cleaner `panic!("{}", err)` -> `panic!("{err}")`

nealsid replied:
Done.

tgross35 commented on library/core/src/cell.rs:
```diff
@@ -806,22 +804,22 @@ fn panic_already_mutably_borrowed(err: BorrowError) -> ! {
-type BorrowFlag = isize;
-const UNUSED: BorrowFlag = 0;
+type BorrowCounter = isize;
+const UNUSED: BorrowCounter = 0;
```
+1 to this change, flag doesn't really make sense here. I was curious where the name came from, looks like it's been around since before Rust 1.0 https://github.com/rust-lang/rust/pull/10514.

nealsid replied:
Thanks!

tgross35 reviewed (commented):
Couple small requests then I think we can get this in!

Comment `@rustbot review` when it's ready for a review (so I don't miss it in my queue)

nealsid commented:
@rustbot review

tgross35 commented:
Thank you!

@bors r+ rollup
